### ISSUE TITLE
Collect new SymbolKinds

### ISSUE THREAD

**Issue by dbaeumer** [Microsoft/language-server-protocol#344]
Collects new symbol kinds.

**Comment by vladdu** [comment]
I assume it's best to propose one item per comment, so that people can vote for it.

**Comment by vladdu** [comment]
- pragma / directive

**Comment by vladdu** [comment]
- macro

**Comment by rictic** [comment]
Slot, in the sense of a named, declared hole in a template that a user of the element can use to insert content.

e.g. `<slot>` in Shadow Dom v1, `<content>` in Shadow Dom v0, transclusion points in AngularJS, `<ng-content>` in Angular.

**Comment by laughedelic** [comment]
There is `CompletionItemKind`, but no `SymbolKind` for  `Value`

**Comment by tsmaeder** [comment]
Annotation

**Comment by raja-s** [comment]
Some languages make a fundamental distinction between a function that has a return type and a "function" that doesn't return anything, unlike in C for example where it's simply determined by whether or not the return type is `void`, so it would be useful to have a separate symbol kind that could be called for example `Procedure`.

**Comment by raja-s** [comment]
`Type` for type declarations.

**Comment by raja-s** [comment]
`Parameter`/`Argument` to be able to distinguish function/constructor/... parameters from actual variables.

**Comment by dbaeumer** [comment]
@kenkangxgwe can you provide me with an example.

**Comment by kenkangxgwe** [comment]
@dbaeumer 
In LaTeX,
```latex
\section{Introduction}
%...
```
In Wolfram Language
```mathematica
(* ::Section:: *)
(*Introduction*)
```

In MATLAB
```matlab
%%Introduction
// ...
```

In C#
```C#
#region section
// ...
#endregion  
```

**Comment by movie-travel-code** [comment]
Should provide a symbol kind for the `type alias`?

go code as bellow.
```go
type DogName = string
```

swift code as bellow.
```swift
typealias DogName = String
```

**Comment by anton-dessiatov** [comment]
That's pretty much Haskell-specific, but we could have a ```pattern``` symbol kind?

There is a GHC Extension named [PatternSynonyms](https://gitlab.haskell.org/ghc/ghc/wikis/pattern-synonyms) that enables syntax for sort-of-first-class patterns like this:

```haskell
data Type = App String [Type]
pattern Arrow t1 t2 = App "->" [t1, t2]
pattern Int = App "Int" []
```

The lack of dedicated symbol kind It's not that big of a problem, currently [Haskell Language Server](https://github.com/haskell/haskell-ide-engine) happily uses `Function` for pattern synonyms, but for the sake of correctness it would be nice to have a special kind.

**Comment by DanTup** [comment]
Dart is adding extension classes, and currently we're having to send SymbolKind=Class to avoid VS code just showing its detail icon. It'd be nice to have some ability to have Extension classes and Extension methods better supported (either as kinds, or modifiers).

There are other comments here suggesting we should just make our own up, but the concerns with that are:

1. Clients will show default icons which may be less appropriate than reusing other icons (for example the VS Code "Class" looks better for extension classes and mixins than the default icon)
2. No guidance is given on the values - if we put them after the existing values, it's possible the spec will gain new ones. Should we just pick arbitrary large numbers?

**Comment by DavidGoldman** [comment]
> Dart is adding extension classes, and currently we're having to send SymbolKind=Class to avoid VS code using a default icon. It'd be nice to have some ability to have Extension classes and Extension methods better supported (either as kinds, or modifiers).
> 
> There are other comments here suggesting we should just make our own up, but the concerns with that are:
> 
> 1. Clients will show default icons which may be less appropriate than reusing other icons (for example the VS Code "Class" looks better for extension classes and mixins than the default icon)
> 2. No guidance is given on the values - if we put them after the existing values, it's possible the spec will gain new ones. Should we just pick arbitrary large numbers?

+1 to adding `extension` support to LSP's `SymbolKind`, potentially with a differentiator between `Interface` extensions and `Class`/`Struct` (implementation) extensions.

Objective-C itself contains [categories](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ProgrammingWithObjectiveC/CustomizingExistingClasses/CustomizingExistingClasses.html) which are widely used:

```
@interface MyClass : NSObject
@property int foo;
@end

@interface MyClass (CategoryAddingNewMethod)
- (void)doSomethingWithFoo;
@end

@implementation MyClass (CategoryAddingNewMethod)
- (void)doSomethingWithFoo {
  NSLog(@"Foo: %@", @(self.foo));
}
@end
```

and Swift contains [extensions](https://docs.swift.org/swift-book/LanguageGuide/Extensions.html), which are also widely used:

```
class MyClass {
  var foo: Int
}
extension MyClass: SomeProtocol {
  func someMethodForProtocol {
    print("hi")
  }
}
```

**Comment by matklad** [comment]
@dbaeumer [`macro`](https://github.com/microsoft/language-server-protocol/issues/344#issuecomment-347202468) has quite a few upvotes, seems pretty uncontroversial to add, and does not really map to any of the existing symbol kinds. 

What's the process for actually adding it to the protocol? I'd love to use it for rust-analyzer

**Comment by dbaeumer** [comment]
@matklad for 3.16 we can make a push to add new kinds. I will synchronize with VS Code as well otherwise they will not be rendered nicely :-)

**Comment by DavidGoldman** [comment]
Friendly ping, is this still on the radar?

**Comment by dbaeumer** [comment]
I checked with the VS Code team and there are currently no plans to add new symbol kinds. To work on this and propose new once we need  a client that implements so. Otherwise only having them in the spec with no implementation is not very helpful.

**Comment by matklad** [comment]
[This list](https://github.com/microsoft/vscode/issues/97063#issuecomment-625182536) (`attribute/annotation`, `union`, `typeAlias`) is what's been implemented in rust-analyzer for a while, and which feels general-purpose enough to uplift to the protocol proper.

**Comment by dbaeumer** [comment]
@matklad this is about `SymbolKind` and not about semantic tokens.

**Comment by matklad** [comment]
Oh, my bad, I confused *both* SemanticTokenType with SymbolKind, and server with client :facepalm: 

But yeah, rust-analyzer will be ready to use `macro`, `typeAlias` and `union` for `SymbolKind` as soon as they are available! I don't think we can do this before the protocol is stabilized though: SymbolKinds, unlike SemanticTokenTypes, are not extensible.

**Comment by sam-mccall** [comment]
> Otherwise only having them in the spec with no implementation is not very helpful

They're valuable once implemented in both server and client, so there's a first-mover disadvantage.
We have several language servers saying they're willing to go first, but are unwilling to break the spec in order to do so. So having them in the spec serves the purpose of coordinating the work (and practically, avoids having servers and other clients blocked on VSCode prioritizing this).

**Comment by dbaeumer** [comment]
IMO `StaticField` and `StaticMethod` shouldn't be new symbol kinds. If we want to express this we should adopt the way how this is handled in semantic tokens. So I would rather have a SymbolModifier which says `static`, `async`, ..

**Comment by KamasamaK** [comment]
@dbaeumer 
> So I would rather have a SymbolModifier which says `static`, `async`, ..

Did you mean `SymbolTag`?

**Comment by dbaeumer** [comment]
@KamasamaK yes sorry. Was confused with semantic tokens.

**Comment by sam-mccall** [comment]
Since this enum can't be extended by clients/servers, it would be useful to have `unknown` or `other` or something. 

For C++ we'd like to be able to include macro *expansions* (not definitions) in the hierarchy in certain cases. (Basically, when a macro dumps a bunch of implementation-detail symbols with into a namespace, the outline is pretty confusing if we can't group those somehow). This is a weird thing to do and I'm not sure it makes sense to add a macroExpansion kind to LSP (especially if VSCode folks don't want more kinds). On the other hand, having to pretend these symbols are nulls or namespaces or something is pretty bad!

(I do agree with your point about StaticField and StaticMethod... the semantic tokens data model is nice!)

**Comment by DavidGoldman** [comment]
Just to confirm, if we wanted to implement this, we'd need to include support for a client as well, likely VS Code, right? How would we go about making contributions to VS Code - what needs to be done to support it in VS Code exactly? New icons + support for the new symbol kinds?

**Comment by dbaeumer** [comment]
If you want new kinds to light up in VS Code then yes, we need support for VS Code as well. We can still extend the list in LSP and guard the new symbol kinds using a capability.

**Comment by gerdleon** [comment]
> Some languages make a fundamental distinction between a function that has a return type and a "function" that doesn't return anything, unlike in C for example where it's simply determined by whether or not the return type is `void`, so it would be useful to have a separate symbol kind that could be called for example `Procedure`.

I would very much appreciate this.

**Comment by gerdleon** [comment]
To support languages that describe database structures `table` would be useful.

**Comment by kraigher** [comment]
I wrote a proposal here https://github.com/microsoft/language-server-protocol/issues/1186#issuecomment-1435928769 that avoids the need to collect a long list symbol kinds globally.

**Comment by yyny** [comment]
`Label`, for C/C++/Zig/JavaScript/Lua/Assembly/LLVM/etc. labels.

**Comment by travkin79** [comment]
`Destructor`

Languages like C/C++ need a distinction between constructor and destructor, so adding `Destructor` to [`SymbolKind`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#symbolKind) would improve the UI (see also CDT LSP issue https://github.com/eclipse-cdt/cdt-lsp/issues/608).

Another option would be to use [`SymbolTags`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#symbolTag) to annotate methods as destructors or constructors, similar to PR #2003 using symbol tags to annotate document symbols with details like public, static, final, and other flags / keywords. In this case I would deprecate `SymbolKind.Constructor` in favour of a new `Constructor` symbol tag (and of course also a `Destructor` symbol tag).
